components/lock.rst

```diff
@@ -0,0 +1,257 @@
+.. index::
+   single: Lock
+   single: Components; Lock
+
+The Lock Component
+==================
+
+    The Lock Component creates and manages `locks`_, a mechanism to provide
+    exclusive access to a shared resource.
+
+.. versionadded:: 3.3
+    The Lock component was introduced in Symfony 3.3.
+
+Installation
+------------
+
+You can install the component in 2 different ways:
+
+* :doc:`Install it via Composer </components/using_components>` (``symfony/lock`` on `Packagist`_);
+* Use the official Git repository (https://github.com/symfony/lock).
+
+.. include:: /components/require_autoload.rst.inc
+
+Usage
+-----
+
+Locks are used to guarantee exclusive access to some shared resource. In
+Symfony applications, you can use locks for example to ensure that a command is
+not executed more than once at the same time (on the same or different servers).
+
+In order to manage the state of locks, you first need to create a ``Store``
+and then use the :class:`Symfony\\Component\\Lock\\Factory` class to actually
+create the lock for some resource::
+
+    use Symfony\Component\Lock\Factory;
+    use Symfony\Component\Lock\Store\SemaphoreStore;
+
+    $store = new SemaphoreStore();
+    $factory = new Factory($store);
+
+Then, call to the :method:`Symfony\\Component\\Lock\\LockInterface::acquire`
+method to try to acquire the lock. Its first argument is an arbitrary string
+that represents the locked resource::
+
+    // ...
+    $lock = $factory->createLock('pdf-invoice-generation');
+
+    if ($lock->acquire()) {
+        // The resource "pdf-invoice-generation" is locked.
+        // You can compute and generate invoice safely here.
+
+        $lock->release();
+    }
+
+If the lock can not be acquired, the method returns ``false``. You can safely
+call the ``acquire()`` method repeatedly, even if you already acquired it.
+
+.. note::
+
+    Unlike other implementations, the Lock Component distinguishes locks
+    instances even when they are created for the same resource. If you want to
+    share a lock in several services, share the ``Lock`` instance returned by
+    the ``Factory::createLock`` method.
+
+Blocking Locks
+--------------
+
+By default, when a lock cannot be acquired, the ``acquire`` method returns
+``false`` immediately. In case you want to wait (indefinitely) until the lock
+can be created, pass ``false`` as the argument of the ``acquire()`` method. This
+is called a **blocking lock** because the execution of your application stops
+until the lock is acquired.
+
+Some of the built-in ``Store`` classes support this feature. When they don't,
+you can decorate them with the ``RetryTillSaveStore`` class::
```

Comment:
"When they don't, you can decorate them with" => "When they don't, they can be decorated with"